### PR TITLE
Adds thumbprint support for JWK

### PR DESCRIPTION
**Pull request by harsimranmaan**
The thumbprint is calculated as described in https://tools.ietf.org/html/rfc7638
This is a breaking changes as it alters the definition of the kid.

### CODE REVIEW

**Comment by harsimranmaan** [comment]
I was expecting a CI build to see if this breaks any tests. I haven't run the test suite. This is a stepping stone for #21

**Comment by harsimranmaan** [comment]
A lot of tests fail on node 4.2.1 but I think that's mostly because of the engine. Lemme know how it goes with the supported engine versions

**Comment by linuxwolf** [comment]
Unfortunately, this patch is not acceptable.  It must not use the "crypto" module directly, for compatibility reasons.  It must also use Promises, again for compatibility reasons.

**Comment by linuxwolf** [comment]
@harsimranmaan Unfortunately we do not.  The general rule-of-thumb we've tried very hard to follow is "do work via Promises", but that is a very generic statement.

For this particular case, some refactoring of the JWKFactory flow is probably the first place to start.  I had started on one approach, albeit it did not generate the thumbprint until explicitly requested, but maybe provides a guideline to continue: https://github.com/linuxwolf/node-jose/tree/thumbprint

I'd happily abandon mine if we can work out the kinks on this direction.

**Comment by linuxwolf** [comment]
Closed as functionality is in #40.  If that PR is not enough, re-open this one with updates, but it must support the Promise-based architecture.
